source/topic3.rst

```diff
@@ -208,16 +208,232 @@ Getters and Setters
 toString
 ^^^^^^^^
 
+* You may remember that ``__repr__`` in python was a special method that would return a string version of the object
+    * And, if you called ``print(some_object)``, the ``__repr__`` would automatically get called
+
+* When we wrote our own classes, we ended up overriding the ``__repr__`` we got for free
+    * The default one simply prints out the object name and a memory address, which wasn't very helpful
+    * ``<__main__.Friend object at 0x7f130d9c52e0>``
+    * *Technically*, the behaviour we get for free is inherited from the closest parent class
+        * We will learn more about this later
+
+* An example of a ``__repr__`` for our ``Friend`` class might look something like this
+* We just print out the name of the ``Friend`` and their email
+
+**Python**
+
+.. code-block:: python
+    :linenos:
+
+    # Python --- __repr__
+    def __repr__(self):
+        return self._firstName + " " + self._lastName \
+               + ":\t" + self._email
+
+
+* We can do the same thing in Java, but in java it's called ``toString()``
+* If we do not write our own, the default behaviour is to return a string of the class name and a *hash code*
+    * ``Friend@77459877``
+    * Again, *technically* the behaviour we get for free is inherited
+
+.. code-block:: java
+    :linenos:
+
+    public String toString() {
+        return firstName + " " + lastName + ":\t" + email;
+    }
+
+* And like Python, we can simply print it and ``toString`` will automatically get called
+    * ``System.out.println(aFriend);``
+
+.. warning::
+
+    The idea is that this *returns* a string; we do not want to just print something
 
 
 equals
 ^^^^^^
 
+* We had another special function that we liked to use called ``__eq__``
+* ``__eq__`` let us define what it means for two of the objects to be equivalent
+* It also let us call ``==`` on the objects too
+
+* In Java, we use the ``equals()`` method to define what it means for two of the objects to be equivalent
+* **HOWEVER**, unlike Python, this doesn't hijack the ``==`` operator
+    * ``==`` for the objects is reserved to check if the objects are literally the same object
+    * Same memory address --- it compares the memory addresses
+
+* On the other hand, ``.equals()`` is used to actually compare the content of the objects in some way
+    * We need to think about what it means for two objects of the type to be &equivalent*
+
+* The default behaviour is to is to check if the objects are in the same memory address (``==``)
+    * Similar to ``toString``, the behaviour we get for free is inherited from the closest parent class
+
+
+* So, what should it mean for two ``Friend`` objects to be *equal*?
+* Well, **we** get to define this
+* We can make it whatever we want
+    * Same eye colour?
+    * Same height and weight?
+    * Same number of skin cells?
+
+* But, we need to really think about what makes sense in our context
+    * Let's go with their ``firstName`` and ``lastName`` being equal
```

Comment:
This is fine for learning purposes, but this is a red flag for class design from my experience. This would mean that two _different_ `Friend` instances could be equal -- extreme pain awaits.

Reply:
Added points about better equals